yrs/src/block.rs

```diff
@@ -275,109 +220,126 @@ impl Item {
+            BLOCK_ITEM_DELETED_REF_NUMBER => ItemContent::Deleted(decoder.read_len()),
+            BLOCK_ITEM_JSON_REF_NUMBER => ItemContent::JSON(decoder.read_string().to_owned()),
+            BLOCK_ITEM_BINARY_REF_NUMBER => ItemContent::Binary(decoder.read_buffer().to_owned()),
+            BLOCK_ITEM_STRING_REF_NUMBER => ItemContent::String(decoder.read_string().to_owned()),
+            BLOCK_ITEM_EMBED_REF_NUMBER => ItemContent::Embed(decoder.read_string().to_owned()),
+            BLOCK_ITEM_FORMAT_REF_NUMBER => ItemContent::Format(
+                decoder.read_string().to_owned(),
+                decoder.read_string().to_owned(),
+            ),
+            BLOCK_ITEM_TYPE_REF_NUMBER => {
+                let type_ref = decoder.read_type_ref();
+                let name = if type_ref == types::TYPE_REFS_XML_ELEMENT
+                    || type_ref == types::TYPE_REFS_XML_HOOK
+                {
+                    Some(decoder.read_key().to_owned())
                 } else {
                     None
                 };
-                let innerPtr = types::TypePtr::Id(ptr);
-                let inner = types::Inner::new(innerPtr, name, type_ref);
+                let inner_ptr = types::TypePtr::Id(ptr);
+                let inner = types::Inner::new(inner_ptr, name, type_ref);
                 ItemContent::Type(inner)
             }
-            8 => { // Content Any
-               ItemContent::Any(update_decoder.read_any())
+            BLOCK_ITEM_ANY_REF_NUMBER => {
+                let len = decoder.read_len() as usize;
+                let mut values = Vec::with_capacity(len);
+                let mut i = 0;
+                while i < len {
+                    values.push(decoder.read_any());
+                    i += 1;
+                }
+                ItemContent::Any(values)
             }
-            9 => { // Content Doc
-               ItemContent::Doc(update_decoder.read_string().to_owned(), update_decoder.read_any())
+            BLOCK_ITEM_DOC_REF_NUMBER => {
+                ItemContent::Doc(decoder.read_string().to_owned(), decoder.read_any())
             }
-            _ => { // Unknown
-                panic!("Unknown content type");
+            info => panic!("ItemContent::decode unrecognized info flag: {}", info),
+        }
+    }
+
+    pub(crate) fn splice(&mut self, offset: usize) -> Option<ItemContent> {
+        match self {
+            ItemContent::Any(value) => {
+                let (left, right) = value.split_at(offset);
+                let left = left.to_vec();
+                let right = right.to_vec();
+                *self = ItemContent::Any(left);
+                Some(ItemContent::Any(right))
             }
+            ItemContent::String(string) => {
+                let (left, right) = string.split_at(offset);
+                let left = left.to_string();
+                let right = right.to_string();
+
+                //TODO: do we need that in Rust?
```

Comment:
There were several discussions around this in Yjs. The best solution seems to be to just disallowing splitting surrogate pairs. The Yjs algorithm is not prone to interleaving, so this should be fine.

FYI: However, what we should do - at some point - is to switch to utf16 strings because that is what is used in Yjs. At least the length of a string `item.len()` should be computed based on the utf16 len, otherwise the algorithm will break when complex characters are inserted.